### PR TITLE
feat: allow usage of noir `#[test]` syntax in stdlib

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves 

## Summary\*

Running tests within the Noir stdlib is a little complicated as nargo will not accept it as a regular crate as the stdlib uses features which are disallowed in other crates. This means that in a number of cases we're placing stdlib unit tests inside of the `test_programs` instead of just using the `#[test]` format inside of the stdlib itself.

This is suboptimal as it separates the implementation from the tests in the repository and once https://github.com/noir-lang/noir/pull/4491 is merged, it will be impossible to use this method to test any private functions within the stdlib.

This PR then adds a new integration test which is equivalent to running `nargo test` on the stdlib and moves some tests over to the stdlib as an example.

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by TomAFrench** [comment]
Ideally we'd have the rust test code closer to the stdlib but it needs access to the test runner so must be in either `nargo` or `nargo_cli`.

File: tooling/nargo_cli/tests/stdlib-tests.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        entry_path: PathBuf::from("main.nr"),
+        name: "dummy".parse().unwrap(),
+        dependencies: BTreeMap::new(),
+    };
+
+    let (mut context, dummy_crate_id) =
+        prepare_package(&file_manager, &parsed_files, &dummy_package);
+
+    let result = check_crate(&mut context, dummy_crate_id, true, false);
+    report_errors(result, &context.file_manager, true, false).unwrap();
```
Comment (TomAFrench) [inline tooling/nargo_cli/tests/stdlib-tests.rs:34]:
```suggestion
    report_errors(result, &context.file_manager, true, false).expect("Error encountered while compiling standard library");
```

**Review by vezenovm — APPROVED** [review]
Looks good. You could probably also move over the `bounded_vec` tests under `noir_test_success`, but I'm not sure on the best place for those tests as it doesn't use any would-be private functions.

**Comment by TomAFrench** [comment]
Yep, I wanted to do this separately however as it's going to cause ripple effects into the docs.

**Review by jfecher — APPROVED** [review]

**Pull request merged**